src/Common/src/System/Security/Cryptography/Asn1V2.cs

```diff
@@ -0,0 +1,5246 @@
+                return true;
+            }
+
+            byte firstByte = (byte)(_controlFlags | TagNumberMask);
+            destination[0] = firstByte;
+
+            int remaining = TagValue;
+            int idx = spaceRequired - 1;
+
+            while (remaining > 0)
+            {
+                int segment = remaining & 0x7F;
+
+                // The last byte doesn't get the marker, which we write first.
+                if (remaining != TagValue)
+                {
+                    segment |= 0x80;
+                }
+
+                Debug.Assert(segment <= byte.MaxValue);
+                destination[idx] = (byte)segment;
+                remaining >>= 7;
+                idx--;
+            }
+
+            Debug.Assert(idx == 0);
+            bytesWritten = spaceRequired;
+            return true;
+        }
+
+        public bool Equals(Asn1Tag other)
+        {
+            return _controlFlags == other._controlFlags && _tagValue == other._tagValue;
+        }
+
+        public override bool Equals(object obj)
+        {
+            if (ReferenceEquals(null, obj)) return false;
+            return obj is Asn1Tag && Equals((Asn1Tag)obj);
+        }
+
+        public override int GetHashCode()
+        {
+            // Most TagValue values will be in the 0-30 range,
+            // the GetHashCode value only has collisions when TagValue is
+            // between 2^29 and uint.MaxValue
+            return (_controlFlags << 24) ^ _tagValue;
+        }
+
+        public static bool operator ==(Asn1Tag left, Asn1Tag right)
+        {
+            return left.Equals(right);
+        }
+
+        public static bool operator !=(Asn1Tag left, Asn1Tag right)
+        {
+            return !left.Equals(right);
+        }
+
+        public override string ToString()
+        {
+            const string ConstructedPrefix = "Constructed ";
+            string classAndValue;
+
+            if (TagClass == TagClass.Universal)
+            {
+                classAndValue = ((UniversalTagNumber)TagValue).ToString();
+            }
+            else
+            {
+                classAndValue = TagClass + "-" + TagValue;
+            }
+
+            if (IsConstructed)
+            {
+                return ConstructedPrefix + classAndValue;
+            }
+
+            return classAndValue;
+        }
+    }
+
+    internal enum AsnEncodingRules
+    {
+        BasicEncodingRules,
+        BER = BasicEncodingRules,
+        CanonicalEncodingRules,
+        CER = CanonicalEncodingRules,
+        DistinguishedEncodingRules,
+        DER = DistinguishedEncodingRules,
+    }
+
+    internal class AsnReader
+    {
+        // ITU-T-REC-X.690-201508 sec 9.2
+        internal const int MaxCERSegmentSize = 1000;
+
+        // ITU-T-REC-X.690-201508 sec 8.1.5 says only 0000 is legal.
+        private const int EndOfContentsEncodedLength = 2;
+
+        private static readonly Text.Encoding s_utf8Encoding = new UTF8Encoding(false, true);
```

Comment:
Can these readonly static fields be consolidated with those in `Asn1Reader` and moved to their own utility class?